bpf/lib/ratelimit.h

```diff
@@ -46,14 +74,28 @@ static inline bool ratelimit_check_and_take(struct ratelimit_key *key,
 
 	now = ktime_get_ns();
 
+	if (!key)
+		return false;
+	metrics_key.usage = key->usage;
+	metrics_value = map_lookup_elem(&RATELIMIT_METRICS_MAP, &metrics_key);
+	if (!metrics_value) {
+		new_metrics_value.dropped = 0;
+		metrics_value = &new_metrics_value;
+		ret = map_update_elem(&RATELIMIT_METRICS_MAP, &metrics_key, metrics_value, BPF_ANY);
+		if (unlikely(ret < 0 || !metrics_value))
+			return false;
+	}
+
 	/* Create a new bucket if we do not yet have one for the key */
 	value = map_lookup_elem(&RATELIMIT_MAP, key);
 	if (!value) {
 		new_value.last_topup = now;
 		new_value.tokens = settings->tokens_per_topup - 1;
 		ret = map_update_elem(&RATELIMIT_MAP, key, &new_value, BPF_ANY);
-		if (ret < 0)
+		if (unlikely(ret < 0)) {
+			metrics_value->dropped++;
```

Comment:
Yeah, it is racy, I agree. There is a comment inside this function that already states that this rate limiter implementation allows inaccuracies resulting from to race conditions, because the alternative would have a rather significant impact on the performance. I can move the comment up a bit, so that this code will be in the scope of that comment as well.

Reply:
Added some additional comment explaining that.